crates/jstz_core/src/kv/outbox.rs

```diff
@@ -0,0 +1,461 @@
+        message
+            .bin_write(&mut buffer)
+            .map_err(|_| OutboxError::OutboxMessageSerializationError)?;
+        rt.write_output(&buffer)?;
+        Ok(())
+    }
+
+    pub fn new(max: u32) -> Self {
+        Self {
+            meta: OutboxQueueMeta { queue_len: 0, max },
+            rollup_outbox_queue: RollupOutboxQueue::new(&ROLLUP_OUTBOX_QUEUE_ROOT, max)
+                .unwrap(),
+        }
+    }
+
+    pub fn queue_len(&self) -> u32 {
+        self.meta.queue_len
+    }
+
+    pub fn incr_queue_len(&mut self) {
+        self.meta.queue_len += 1;
+    }
+
+    pub fn max(&self) -> u32 {
+        self.meta.max
+    }
+
+    /// Flushes the outbox queue in the order of the rollup outbox queue
+    /// then the outbox queue snapshot. The outbox has a maximum capacity
+    /// of 100 messages per level. Messages that cannot be flushed in the
+    /// current level are enqueued into the rollup outbox queue for the
+    /// next flush.
+    pub fn flush(
+        &mut self,
+        host: &mut impl Runtime,
+        snapshot: OutboxQueueSnapshot,
+    ) -> usize {
+        // 1. Flush the existing outbox queue
+        let mut flushed_count = self.rollup_outbox_queue.flush_queue(host);
+
+        // 2. Flush the outbox queue snapshot if there is space in the outbox
+        let mut tezos_outbox_messages =
+            snapshot.0.into_iter().map(|message| message.into());
+
+        for message in tezos_outbox_messages.by_ref() {
+            match Self::write_outbox_message(host, &message) {
+                Ok(()) => {
+                    flushed_count += 1;
+                    continue;
+                }
+                Err(crate::Error::HostError {
+                    source:
+                        tezos_smart_rollup::host::RuntimeError::HostErr(
+                            tezos_smart_rollup_host::Error::FullOutbox,
+                        ),
+                }) => {
+                    self.rollup_outbox_queue
+                        .queue_message(host, message)
+                        .unwrap();
+                    break;
+                }
+                Err(e) => {
+                    // This arm is unexpected and probably indicates a bug
+                    // or cpu/memory degradation.
+                    debug_msg!(host, "Error while writing message to outbox: {:?}", e);
+                    self.rollup_outbox_queue
+                        .queue_message(host, message)
+                        .unwrap();
+                    break;
+                }
+            }
+        }
+
+        //  3. Enqueue the remaining messages into the outbox queue
+        for message in tezos_outbox_messages {
+            self.rollup_outbox_queue
+                .queue_message(host, message)
+                .expect("Should always be able to queue message");
+        }
+
+        // 4. Finally, update the counter outbox messages
+        self.meta.queue_len -= flushed_count as u32;
```

Comment:
Incorrect `queue_len` since we don't update `self.meta` when calling `rollup_outbox_queue.queue_message`

Reply:
Ah, we don't have to. `self.meta.queue_len` is the sum of `rollup_outbox_queue` length and `OutboxQueueSnapshot` length. We use `rollup_outbox_queue.queue_message` here to simply move messages from in-memory queue to the persistent queue.

Reply:
That seems a rather messy/error prone invariant -- since `OutboxQueueSnapshot` isn't contained within `OutboxQueue`? What I suggest is that a `Transaction` has a notion of a 'combined' outbox queue length (since it contains both the list of snapshots to write and the persistent outbox queue).

Reply:
I moved the meta into `Transaction` and added RollupOutboxQueue to Transaction. `meta` tracks the combined queue length of snapshots and rollup outbox queues and ensures the queue length is limited to `max` when messages are enqueued in the snapshot. 

`flush` no longer does the queue len accounting.Iinstead, the accounting is done by `Transaction` since it holds all the information - `outbox_queue_meta`, `rollup_outbox_queue` and  `snapshop.snapshot_outbox_queue`

Reply:
Btw I pushed the refactoring as second commit so that you can see the changes or revert if needed. Will squash eventually.

Reply:
I'm still not entirely satisfied with this separation + it leaks the invariant still (via `default`). My take is that we should have two structs:
- `PersistentOutboxQueue`
- `SnapshotOutboxQueue`

The total length of the non-persistent (snapshot) outbox queues is maintained by a field in `Transaction` maintained by `Transaction::queue_outbox_message`. 

The maximum and persistent length is managed by `PersistentOutboxQueue` using a `queue_meta : Option<OutboxQueueMeta>` field. This permits us to lazily read the metadata by passing a `&rt` in getters (`PersistentOutboxQueue::queue_len(&impl Runtime)` and `PersistentOutboxQueue::max_len(&impl Runtime)`)
Flushing will decrease the in-memory representation of the persistent length. 

The transactions total outbox queue length is simply the sum of the two. `Transaction::queue_outbox_message` is modified to take an `&rt`.

Reply:
Made the changes as discussed here and on call